### DISCUSSION TITLE
[solved] Why does the VPN server has to have the root certificat in x509ca

### DISCUSSION THREAD

**Discussion by robertdahlem** [strongswan/strongswan#567] [General]
Hello,

I have been having issues with `no issuer certificate found for "CN=vpn.myserver.com"` on the client side. It turned out that I did not have the Let's Encrypt root certificate in x509ca.

From [Issue #2812 Note 5](https://wiki.strongswan.org/issues/2812#note-5):

> For authentication to be successful, at least the server side has to have the complete chain from the root certificate to its own certificate

It says it, I configured it and now it works.

What I don't understand: **why does a server has to have the root certificate**? What I thought before: the server presents its certificate and - if necessary - a chain certificate. The client has to have the root certificate because it should not trust some random Root CA from the internet.

Regards,
Robert

**Comment by Thermi** [comment]
The server needs to be able to be able to authenticate the, for example, OCSP or CRL responder of the issuer CA of its own certificate in order to fetch those responses and provide them to IKE peers.

Potentially for finding the CA mapped to by a CERT request (SHA-1'd issuer DN AFAIR).

**Reply by tobiasbrunner** [reply]
>  OCSP or CRL responder of the issuer CA of its own certificate in order to fetch those responses and provide them to IKE peers

No, the client fetches those on its own. strongSwan does not send CRLs in CERT payloads and RFC 4806 isn't supported either (and even so it would still be up to the client to verify those). strongSwan does no online verification of its own certificate.

**Reply by Thermi** [reply]
The reply was meant generelly, as the question was specifically for A server, not specifically for strongSwan.
HTTP servers frequently use OCSP and CRL stapling, sometimes with Certificate Transparency replies too.

**Comment by tobiasbrunner** [comment]
> why does a server has to have the root certificate?

It doesn't necessarily. But as @Thermi mentioned, it could make a difference in regards to certificate requests (see e.g. https://github.com/strongswan/strongswan/discussions/473#discussioncomment-1164690).

**Reply by robertdahlem** [reply]
> It doesn't necessarily.

I'm confused now :-) because that doesn't explain `no issuer certificate found for "CN=vpn.myserver.com"` without the root certificate or the absence of that error message when the server has its root certificate.

Instead I would have expected this to happen:

> If you didn't have the root CA certificate installed on the server, the intermediate CA certificate would actually have been sent as there would not have been any known trust anchor and the incomplete trust chain would get accepted.

I even had the intermediate CA added in `connections.<conn>.local<suffix>.certs` (stored under `x509`) but that didn't help either.

**Reply by tobiasbrunner** [reply]
> Instead I would have expected this to happen:

Read the logs to see what's actually happening if you don't load the root CA certificate.

> I even had the intermediate CA added in `connections.<conn>.local<suffix>.certs` (stored under `x509`) but that didn't help either.

Both that option and that directory are for end-entity certificates. Intermediate CA certificates should be stored in `x509ca` (and could be referenced in `cacerts` or `ca_id`, but those are more for client certificate validation).

**Reply by robertdahlem** [reply]
> Read the logs to see what's actually happening if you don't load the root CA certificate.

When the root CA is not loaded, the second line of the following is missing:
`sending end entity cert "CN=vpn.myserver.com"`
`sending issuer cert "C=US, O=Let's Encrypt, CN=R3"`
which baffles me: it is the intermediate CA (which BTW is loaded from `x509`, not from `x509ca`) and according to `swanctl --list-certs` is loaded in both cases.

So, when the root CA is not loaded, the server does not send the intermediate CA. That stands in contrast to
> If you didn't have the root CA certificate installed on the server, the intermediate CA certificate would actually have been sent as there would not have been any known trust anchor and the incomplete trust chain would get accepted.

(I tried having the intermediate both in  `x509` and `x509ca`).

Sorry for being that persistent. I just would like to understand what's going on.

**Reply by tobiasbrunner** [reply]
> When the root CA is not loaded, the second line of the following is missing:

Why didn't you post the complete log? In particular the loaded certificates and the certificate requests would be interesting.

> (which BTW is loaded from `x509`, not from `x509ca`)

Yeah, CA certificates are loaded fine from the first directory (the daemon doesn't care where a certificate was stored on the vici level, although you won't be able to load an end-entity certificate from `x509ca` where only certificates with CA basic constraint are accepted), but you won't be able to reference them in `cacerts` (that's evaluated by swanctl and only works with the `x509ca` directory).

> according to swanctl --list-certs is loaded in both cases

That output would be of interest too.

> That stands in contrast to

It does. And since this works perfectly fine here (i.e. not having the root installed and still sending the intermediate CA certificate) I don't yet know what's going wrong in your case. What version are you using?

**Reply by robertdahlem** [reply]
> Why didn't you post the complete log?

Because I thought this would be an easy one. :-)

Here we go: good.* is with `ISRG_Root_X1.pem` in `x509ca`, bad.* is without that.

Server version is 5.7.2-1 under Debian 10. Client version is 5.9.3rc1 under Android 10.
Both connection attempts made after a fresh restart of `strongswan-swanctl.service`.

[bad.client.log](https://github.com/strongswan/strongswan/files/7084740/bad.client.log)
[bad.server.log](https://github.com/strongswan/strongswan/files/7084741/bad.server.log)
[bad.list-certs.txt](https://github.com/strongswan/strongswan/files/7084742/bad.list-certs.txt)
[good.client.log](https://github.com/strongswan/strongswan/files/7084743/good.client.log)
[good.server.log](https://github.com/strongswan/strongswan/files/7084744/good.server.log)
[good.list-certs.txt](https://github.com/strongswan/strongswan/files/7084745/good.list-certs.txt)

**Reply by robertdahlem** [reply]
Oops, I forgot this:
[swanctl.conf.txt](https://github.com/strongswan/strongswan/files/7090151/swanctl.conf.txt)

**Reply by tobiasbrunner** [reply]
I see, it's basically the same issue discussed in #473. Except that here it's triggered by a cert request for your own custom CA that you use for client certificates. It acts as trust anchor, but there is obviously no path from the server cert to it, so the incomplete trust chain is discarded and no intermediate CA certs are sent.

To avoid that, install the root certificate for the server certificate's chain, as you already did, so that it has the complete chain available and the received cert request for it can act as trust anchor.  Alternatively, don't install your custom CA cert on the client (it doesn't need it unless it has to verify other client certs), so no cert request for it will be received.

**Reply by robertdahlem** [reply]
I went with omitting the custom CA cert at the client side (why should the client have that anyway? sort of my original question, but for the client) and now the connection can be established without the server having the Let's Encrypt root certificate. Thank you, Tobias!

I suggest to remove `-certfile caCert.pem` from the `openssl pkcs12` example in [Setting-up a Simple CA Using the strongSwan PKI Tool](https://wiki.strongswan.org/projects/strongswan/wiki/SimpleCA) because that's what led me to this. :-)

**Reply by tobiasbrunner** [reply]
I've added a note. In the how-to scenario, the server certificate is actually issued by the same custom CA, so installing the CA certificate on the client via PKCS#12 makes perfect sense.
